File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.java
import java.util.*;
class Solution {
    public int solution(int a, int b, int c, int d) {
        int answer =0;
        int dice [] = {a,b,c,d};
        Arrays.sort(dice);
        
        Map<Integer,Integer>map = new HashMap<Integer,Integer>();

        for(int i=0; i<dice.length; i++){
            if(!map.containsKey(dice[i])){
                map.put(dice[i],map.getOrDefault(dice[i], 0) + 1);
            }else{
                map.put(dice[i],map.getOrDefault(dice[i], 0) + 1);
            }
        }
            System.out.print(map);
        if(map.size() == 1) answer = 1111 * dice[0];
        if(map.size() == 4) answer = dice[0];
        
        if(map.size() == 2){
            if(dice[1] == dice[2]){
                if(dice[0] != dice[1]){
                    answer = (int)(Math.pow((10 * dice[1] + dice[0]),2));
                }else{
                    answer = (int)(Math.pow((10 * dice[2] + dice[3]),2));
                }
            }else{
                answer = (dice[1] + dice[2]) * Math.abs(dice[1] - dice[2]);
            }
        } 
        if(map.size() == 3){
            for(int i=0; i<dice.length-1; i++){
                if(dice[i] == dice[i+1]){
                    if(i == 0) answer = dice[2] * dice[3];
                    if(i == 1) answer = dice[0] * dice[3];
                    if(i == 2) answer = dice[0] * dice[1];
                }
            }
            }
        System.out.print(dice[0]+" "+dice[1]+" "+dice[2]+" "+dice[3]);
        System.out.print(Math.pow((10 * dice[0] + dice[3]),2));    
        
        return answer;
    }
}

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.js
function solution(a, b, c, d) {
    let dice = [a,b,c,d];
    let count = {};
    
    dice.forEach((num)=>{
        count[num] = (count[num] || 0) + 1;
    });
    
    const keys = Object.keys(count).map(Number).sort((a,b)=>a-b);
    
    if(keys.length === 1){
        return keys[0] * 1111;
    }
    
    if(keys.length === 2){
        const p = keys[0];
        const q = keys[1];
        if(count[p] === 3){
            return Math.pow(10* p + q ,2);
        }else if(count[q] === 3){
            return Math.pow(10* q + p ,2)
        }else{
            return (p + q) * Math.abs(p - q);
        }
    }
    
    if(keys.length === 3){
        const p = keys[0];
        const q = keys[1];
        const r = keys[2];
       if(count[p] === 2){
            return q*r;
        }else if(count[q] === 2){
            return p*r;
        }else if(count[r] === 2){
            return p*q;
        }
    }
    
   if(keys.length === 4){
       return keys[0];
   }
}